src/rgw/rgw_sal.cc

```diff
@@ -166,4 +166,200 @@ int Object::range_to_ofs(uint64_t obj_size, int64_t &ofs, int64_t &end)
   }
   return 0;
 }
+
+std::unique_ptr<rgw::sal::Object> MultipartUpload::create_meta_obj(rgw::sal::Store *store, RGWMPObj& mp_obj)
+{
+  std::string oid = mp_obj.get_key();
+
+  char buf[33];
+  string tmp_obj_name;
+  std::unique_ptr<rgw::sal::Object> obj;
+  gen_rand_alphanumeric(store->ctx(), buf, sizeof(buf) - 1);
+  std::string upload_id = MULTIPART_UPLOAD_ID_PREFIX; /* v2 upload id */
+  upload_id.append(buf);
+
+  mp_obj.init(oid, upload_id);
+  tmp_obj_name = mp_obj.get_meta();
+
+  obj = get_meta_obj();
+
+  return obj;
+}
+
+int MultipartUpload::get_sorted_omap_part_str(int marker, std::string& p) {
+    p = "part.";
+    char buf[32];
+
+    snprintf(buf, sizeof(buf), "%08d", marker);
+    p.append(buf);
+    return 0;
+}
+
+int MultipartUpload::list_parts(const DoutPrefixProvider *dpp, CephContext *cct,
+				     int num_parts, int marker,
+				     int *next_marker, bool *truncated,
+				     bool assume_unsorted)
+{
+  map<string, bufferlist> parts_map;
+  map<string, bufferlist>::iterator iter;
+
+  std::unique_ptr<rgw::sal::Object> obj = get_meta_obj();
+  obj->set_in_extra_data(true);
+
+  bool sorted_omap = is_v2_upload_id(get_upload_id()) && !assume_unsorted;
+
+  parts.clear();
+
+  int ret;
+  if (sorted_omap) {
+    std::string p;
+    (void)get_sorted_omap_part_str(marker, p);
+
+    ret = obj->omap_get_vals(dpp, p, num_parts + 1, &parts_map,
+                                 nullptr, null_yield);
```

Comment:
i'm confused by the use of omap in the generic `MultipartUpload` class. omap and this is_v2_upload_id stuff should be implementation details of the `RadosMultipartUpload`

do we really intend to keep these omap interfaces in `rgw::sal::Object`, or are are they just temporary?

i would expect dbstore to implement this list of multipart parts with a table

Reply:
omap_*_val*() APIs have been part of of rgw::sal::Object since the beginning. I am not sure if they are temporary. @dang can confirm it.

If they need to be removed from common sal, I shall add another routine to fetch parts_map() from the backend here.

Reply:
I always assumed that OMAP was part of the API and would always be there.  The API is currently only used by multipart and torrent, so it could possibly be removed, if we add a specific API for torrent.

Reply:
ok, we can leave this for now until further discussion. i'll share my thoughts in the meantime:

i don't think omap really makes sense for `rgw::sal::RadosObject` either. omap isn't supported on EC pools, so we can't use it on head objects in general. that was an oversight with the torrent implementation (see https://tracker.ceph.com/issues/23838), which should be using an xattr instead of omap. for multipart uploads, we've had to put these 'meta' objects in a separate non-ec pool as a workaround

i really don't think `rgw::sal::Object` is a good fit for this multipart 'meta' object. it's not an object visible to s3, and very little of `rgw::sal::Object`'s interface actually makes sense there. we also shouldn't have to expand the `rgw::sal::Object` interface with things like `bool in_extra_data` that don't apply to normal objects

it's fine for backends to share some code between s3 objects and the multipart meta object; for example, the rados backend needs to store both in the bucket index. but that doesn't mean they should share a common sal interface

Reply:
Instead of omap, I used list<parts>  now (part of the meta object head) to store and list parts info, during multipart upload.  Also have reverted the common sal interfaces added earlier. Please review.

Reply:
@dang @cbodley ..request to review the latest changes and let me know if you have any further comments. Thanks!